spotipy/oauth2.py

```diff
@@ -505,10 +506,248 @@ def _add_custom_values_to_token_info(self, token_info):
+        * scope: May be supplied, intuitively converted to proper format
+        * cache_path: May be supplied, will otherwise be generated
+        * username: Must be supplied or set as environment variable
+        * show_dialog: Interpreted as boolean
+        """
+        self.client_id = client_id
+        self.redirect_uri = redirect_uri
+        self.state = state
+        self.cache_path = cache_path
+        self.username = username or os.getenv(
+            CLIENT_CREDS_ENV_VARS["client_username"]
+        )
+        self.scope = self._normalize_scope(scope)
+        self.show_dialog = show_dialog
+        self._session = None  # As to not break inherited __del__
+
+    def get_cached_token(self):
+        """ Gets a cached auth token
+        """
+        token_info = None
+
+        if not self.cache_path and self.username:
+            self.cache_path = ".cache-" + str(self.username)
+        elif not self.cache_path and not self.username:
+            raise SpotifyOauthError(
+                "You must either set a cache_path or a username."
+            )
+
+        if self.cache_path:
+            try:
+                f = open(self.cache_path)
+                token_info_string = f.read()
+                f.close()
+                token_info = json.loads(token_info_string)
+
+                # if scopes don't match, then bail
+                if "scope" not in token_info or not self._is_scope_subset(
+                    self.scope, token_info["scope"]
+                ):
+                    return None
+
+                if self.is_token_expired(token_info):
+                    return None
+
+            except IOError:
+                pass
+        return token_info
+
+    def _save_token_info(self, token_info):
+        if not self.cache_path and self.username:
+            self.cache_path = ".cache-" + str(self.username)
+        if self.cache_path:
+            try:
+                f = open(self.cache_path, "w")
+                f.write(json.dumps(token_info))
+                f.close()
+            except IOError:
+                logger.warning('Couldn\'t write token to cache at: %s',
+                               self.cache_path)
+
+    def _is_scope_subset(self, needle_scope, haystack_scope):
+        needle_scope = set(needle_scope.split()) if needle_scope else set()
+        haystack_scope = (
+            set(haystack_scope.split()) if haystack_scope else set()
+        )
+        return needle_scope <= haystack_scope
+
+    def is_token_expired(self, token_info):
+        return is_token_expired(token_info)
+
+    def get_access_token(self,
+                         as_dict=False,
+                         state=None,
+                         response=None,
+                         check_cache=True):
+        """ Gets Auth Token from cache (preferred) or user interaction
+
+        Parameters
+        ----------
+        * as_dict: **DO NOT USE**, included for tests compatability
```

Comment:
I can't find what is using this in the tests, could we not just delete it? I would rather not keep stuff that's only needed in tests.

compatability -> compatibility

Reply:
It was for `prompt_for_user_token`. Once `prompt_for_user_token` was deprecated, I switched to the new template, so it should be removable.